ssh2/libssh2_enums.pyx

```diff
@@ -0,0 +1,194 @@
+    CRYPT_SC = c_ssh2.LIBSSH2_METHOD_CRYPT_SC
+    MAC_CS = c_ssh2.LIBSSH2_METHOD_MAC_CS
+    MAC_SC = c_ssh2.LIBSSH2_METHOD_MAC_SC
+    COMP_CS = c_ssh2.LIBSSH2_METHOD_COMP_CS
+    COMP_SC = c_ssh2.LIBSSH2_METHOD_COMP_SC
+    LANG_CS = c_ssh2.LIBSSH2_METHOD_LANG_CS
+    LANG_SC = c_ssh2.LIBSSH2_METHOD_LANG_SC
+
+
+class Flag(enum.Enum):
+    SIGPIPE = c_ssh2.LIBSSH2_FLAG_SIGPIPE
+    COMPRESS = c_ssh2.LIBSSH2_FLAG_COMPRESS
+
+
+class SFTP(enum.Enum):
+    # Type of file mask
+    S_IFMT = c_sftp.LIBSSH2_SFTP_S_IFMT
+    # named pipe (fifo)
+    S_IFIFO = c_sftp.LIBSSH2_SFTP_S_IFIFO
+    # character special
+    S_IFCHR = c_sftp.LIBSSH2_SFTP_S_IFCHR
+    # directory
+    S_IFDIR = c_sftp.LIBSSH2_SFTP_S_IFDIR
+    # block special (block device)
+    S_IFBLK = c_sftp.LIBSSH2_SFTP_S_IFBLK
+    # regular
+    S_IFREG = c_sftp.LIBSSH2_SFTP_S_IFREG
+    # symbolic link
+    S_IFLNK = c_sftp.LIBSSH2_SFTP_S_IFLNK
+    # socket
+    S_IFSOCK = c_sftp.LIBSSH2_SFTP_S_IFSOCK
+
+    # File Transfer Flags
+    FXF_READ = c_sftp.LIBSSH2_FXF_READ
+    FXF_WRITE = c_sftp.LIBSSH2_FXF_WRITE
+    FXF_APPEND = c_sftp.LIBSSH2_FXF_APPEND
+    FXF_CREAT = c_sftp.LIBSSH2_FXF_CREAT
+    FXF_TRUNC = c_sftp.LIBSSH2_FXF_TRUNC
+    FXF_EXCL = c_sftp.LIBSSH2_FXF_EXCL
+
+    # File mode masks
+    # Read, write, execute/search by owner
+    S_IRWXU = c_sftp.LIBSSH2_SFTP_S_IRWXU
+    S_IRUSR = c_sftp.LIBSSH2_SFTP_S_IRUSR
+    S_IWUSR = c_sftp.LIBSSH2_SFTP_S_IWUSR
+    S_IXUSR = c_sftp.LIBSSH2_SFTP_S_IXUSR
+    # Read, write, execute/search by group
+    S_IRWXG = c_sftp.LIBSSH2_SFTP_S_IRWXG
+    S_IRGRP = c_sftp.LIBSSH2_SFTP_S_IRGRP
+    S_IWGRP = c_sftp.LIBSSH2_SFTP_S_IWGRP
+    S_IXGRP = c_sftp.LIBSSH2_SFTP_S_IXGRP
+    # Read, write, execute/search by others
+    S_IRWXO = c_sftp.LIBSSH2_SFTP_S_IRWXO
+    S_IROTH = c_sftp.LIBSSH2_SFTP_S_IROTH
+    S_IWOTH = c_sftp.LIBSSH2_SFTP_S_IWOTH
+    S_IXOTH = c_sftp.LIBSSH2_SFTP_S_IXOTH
+
+    # Read only
+    ST_RDONLY = c_sftp.LIBSSH2_SFTP_ST_RDONLY
+    # No suid
+    ST_NOSUID = c_sftp.LIBSSH2_SFTP_ST_NOSUID
+
+
+class KnownHost(enum.Enum):
+    # Host format type masks
+    TYPE_MASK = c_ssh2.LIBSSH2_KNOWNHOST_TYPE_MASK
+    TYPE_PLAIN = c_ssh2.LIBSSH2_KNOWNHOST_TYPE_PLAIN
+    TYPE_SHA1 = c_ssh2.LIBSSH2_KNOWNHOST_TYPE_SHA1
+    TYPE_CUSTOM = c_ssh2.LIBSSH2_KNOWNHOST_TYPE_CUSTOM
+
+    # Key format type masks
+    KEYENC_MASK = c_ssh2.LIBSSH2_KNOWNHOST_KEYENC_MASK
+    KEYENC_RAW = c_ssh2.LIBSSH2_KNOWNHOST_KEYENC_RAW
+    KEYENC_BASE64 = c_ssh2.LIBSSH2_KNOWNHOST_KEYENC_BASE64
+
+    # Key type masks
+    KEY_MASK = c_ssh2.LIBSSH2_KNOWNHOST_KEY_MASK
+    KEY_SHIFT = c_ssh2.LIBSSH2_KNOWNHOST_KEY_SHIFT
+    KEY_RSA1 = c_ssh2.LIBSSH2_KNOWNHOST_KEY_RSA1
+    KEY_SSHRSA = c_ssh2.LIBSSH2_KNOWNHOST_KEY_SSHRSA
+    KEY_SSHDSS = c_ssh2.LIBSSH2_KNOWNHOST_KEY_SSHDSS
+    IF EMBEDDED_LIB:
+        KEY_UNKNOWN = c_ssh2.LIBSSH2_KNOWNHOST_KEY_UNKNOWN
+
+
+class ErrorCodes(enum.Enum):
+    NONE = error_codes._LIBSSH2_ERROR_NONE
```

Comment:
These can already be imported in python `from ssh2.error_codes import LIBSSH2_ERROR_NONE`.

If this way of using them is easier, the existing definitions would need updating, not to duplicate them.

Reply:
Using enums for imports is a different issue in line with something more like [this](https://github.com/Red-M/Redlibssh2/blob/master/ssh2/__init__.py#L5-L18) as importing select pieces, particularly just a single attribute import makes downstream code like [this](https://github.com/Red-M/RedSSH/blob/master/redssh/libssh2.py#L20-L28). Which is horrible to read and understand what is being imported due to just the sheer number of objects being imported. Not to mention global imports are easier to deal with to being with.

The enums is one thing but there is some attributes that are missing that you'll see in the other files.

These definitions are not duplicates but these can be done either this way or by having the rest of the library use them and this was an easier way to implement enums to allow for ease of use and importing in general. I couldn't use these because I didn't have package builds for [my fork](https://github.com/Red-M/Redlibssh2) up yet so I had to make patches in my fork and do workarounds in my downstream library in the hopes you'd merge the changes in.

Reply:
Also this maps the C enums to Python enums which makes them "real" enums in Python because the C enums are just values/symbols that can't be type checked, etc.

Reply:
Will not getting merged as-is. There are lots of ways to make imports look nicer, can find examples in my other projects.

Adding duplicate definitions of attributes that already exist is not appropriate. If they are to be renamed for whatever reason they would need updating *everywhere they are used in the library*.

I don't see how

```
from ssh2.enums import ErrorCodes
eagain = ErrorCodes.LIBSSH2_ERROR_EAGAIN
```

is better than

```
from ssh2.session import LIBSSH2_ERROR_EAGAIN
```

Most of what has been added here already exists in `ssh2.session`, `ssh2.sftp` and elsewhere and are not necessary changes.

The C enums are not named in libssh2, which is why they were not put into named enums in the first place. 

The Python attributes in ssh2.session ssh2.error_codes map to their C counter parts and can be used directly. Suggest to add what is missing any functionality that needs it.

Reply:
These aren't duplicates but if you're wanting to reject based on that and assumptions, I don't think taking patches from me is going to work out.